### PR TITLE
Stream pipe API - fixes #2792

### LINKED ISSUE #2792: Stream pipe API
Currently the advocated way to transfer data between streams is to use the pump API. This API only cares about transferring data and leave the aspects of stream life cycle and error management to the developer, which can be difficult to implement when a write stream is obtained asynchronously.

The stream API is a new API that will pipe a read stream to a write stream using the well known pump mechanism and will also take care of other aspects to provide a simple and effective solution to the developer.

The previous pump API will still be supported but not advocated anymore in the doc in favor of the pipe API.

### PR DESCRIPTION
**Pull request by vietj**

### CODE REVIEW

**Comment by vietj** [comment]
the CI does not pass because this PR has a codegen modification to allow this.

**Comment by vietj** [comment]
I agree on everything you said @jponge we'll merge this after 3.6.3 and promote 3.6 to 3.7

**Comment by jponge** [comment]
👍 👍 👍 👍

**Comment by azagniotov** [comment]
@vietj could you provide a rough estimate when this feature is going to be available live? How is your timeline looking?

**Comment by vietj** [comment]
@azagniotov this is target for 3.7 in february / march

**Comment by jponge** [comment]
@vietj It may be useful to put the timelines for 3.6.3 / 3.7 and 4.0 at https://github.com/vert-x3/wiki/wiki/Vert.x-Roadmap

**Comment by vietj** [comment]
sure thing

**Comment by jponge** [comment]
Looks like you have a CI failure

**Comment by vietj** [comment]
yes the project does not compile because of missing codegen

**Comment by vietj** [comment]
actually now it has the relevant commits

**Comment by jponge** [comment]
Thanks I'm reviewing soon

File: src/main/asciidoc/streams.adoc [hunk 1]
```diff
@@ -125,6 +164,17 @@ The handler will be called if the `WriteStream` is considered no longer full.
...
 
+The pump exposes a subset of the pipe API and only transfers the items between streams, it does not handle
+the completion or failure of the transfer operation.
+
+[source,$lang]
+----
+{@link examples.StreamsExamples#pump(io.vertx.core.Vertx)}
+----
+
+Before Vert.x 3.7 the {@link io.vertx.core.streams.Pump} was the advocated API for transferring a read stream to
```
Comment (jponge) [inline src/main/asciidoc/streams.adoc]: An admonition like `IMPORTANT` would be useful here.

**Comment by vietj** [comment]
The PR now is valid as codegen as been updated with https://github.com/vert-x3/vertx-codegen/issues/239

**Comment by vietj** [comment]
is it OK for merging ?

**Comment by jponge** [comment]
All good for me! Good job mate.


________________________________
De : Julien Viet <notifications@github.com>
Envoyé : lundi, février 11, 2019 11:52 AM
À : eclipse-vertx/vert.x
Cc : Julien Ponge; Mention
Objet : Re: [eclipse-vertx/vert.x] Stream pipe API - fixes #2792 (#2793)


is it OK for merging ?

**Comment by vietj** [comment]
I'll merge this afternoon and port to master

**Pull request merged**